cluster/partitioner/partitioner.go

```diff
@@ -12,41 +18,123 @@ type Partitioner interface {
 }
 
 type Kafka struct {
-	PartitionBy string
-	Partitioner sarama.Partitioner
+	PartitionBy     string
+	Partitioner     sarama.Partitioner
+	GetPartitionKey func(schema.PartitionedMetric, []byte) []byte
 }
 
 func NewKafka(partitionBy string) (*Kafka, error) {
+	kafka := Kafka{
+		PartitionBy: partitionBy,
+	}
+
 	switch partitionBy {
 	case "byOrg":
+		kafka.Partitioner = sarama.NewHashPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyByOrgId(b) }
 	case "bySeries":
+		kafka.Partitioner = sarama.NewHashPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyBySeries(b) }
+	case "bySeriesWithTags":
+		kafka.Partitioner = sarama.NewHashPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyBySeriesWithTags(b) }
 	default:
-		return nil, fmt.Errorf("partitionBy must be one of 'byOrg|bySeries'. got %s", partitionBy)
+		return nil, fmt.Errorf("partitionBy must be one of 'byOrg|bySeries|bySeriesWithTags'. got %s", partitionBy)
 	}
-	return &Kafka{
-		PartitionBy: partitionBy,
-		Partitioner: sarama.NewHashPartitioner(""),
-	}, nil
+
+	return &kafka, nil
 }
 
 func (k *Kafka) Partition(m schema.PartitionedMetric, numPartitions int32) (int32, error) {
-	key, err := k.GetPartitionKey(m, nil)
+	key := k.GetPartitionKey(m, nil)
+	return k.Partitioner.Partition(&sarama.ProducerMessage{Key: sarama.ByteEncoder(key)}, numPartitions)
+}
+
+type jumpPartitionerMauro struct{}
+
+func (p jumpPartitionerMauro) Partition(message *sarama.ProducerMessage, numPartitions int32) (int32, error) {
+	key, err := message.Key.Encode()
```

Comment:
This is not a good idea. Because we compute the partitionIds ourselves, we dont set the message key on kafka messages.

Reply:
I know we've discussed that in the past, but did we actually go forward with that?
seems like we don't do this yet https://github.com/raintank/tsdb-gw/blob/bec070d4ee3e48fa8b4943d040a88ff91cbfce5d/publish/kafka/publish.go#L223-L227

also, this will need a bit more research. It seems to me sarama's topicProducer will always set the msg.Partition via `tp.partitionMessage(msg)` which seems run the Partitioner on the key and seems to ignore a previously set `msg.Partition`.

Reply:
looks like you are right.  But i think we should adopt this approach.  Sarama supports this by setting 
```
config.Producer.Partitioner = sarama.NewManualPartitioner(topic)
```
It will then honour the Partition set in the msg.

Reply:
if we can just set the partitionId and not have to pass the "key" to sarama, we can:

a) cache the partitionId for each metric and not need to compute it for every datapoint.

b) implement the "partitioning" directly in raintank/schema to allow for optimizations and reduced allocations, eg add a `Partition(num int32, p Partitioner) uint32`  to the schema.PartitionedMetric interface, where `Partitioner` is an interface shared by all of our 'partitioner' implementations.  These will likely just need a `Write([]byte)` and `Partition(num int32) uint32`.   Then instead of allocating the "key" to a []byte, and then passing that to a hash function, we can just write directly to the hash, then call sum32/sum64 of the hash func, pass that to jump.Hash() and return the result

Reply:
a) yes, but caching is a concern for the caller/host application, if it makes sense (in this case tsdbgw)
b) clever. but not an obvious home run. This approach implies doing many smaller writes to the hash instead of one. note that with the current implementation we could pool the temporary keys.  So an interesting idea, but taking it seriously requires more R&D time which i don't want to put it now. Let's just get this out and see how it goes. I doubt performance will become an issue.

Reply:
PS: xxhash and go-jump are really fast and use basically no memory. it wouldn't surprise me if adding a cache puts us in worse position

Reply:
if we want to do the partitioning outside of sarama, it means we need the number of partitions, which needs to be queried from the client, which the producer does not expose.
we can however instantiate the client separately, so i'll do that.
one interesting benefit we have is we no longer need to check the list of partitions for each message/produce request. as by convention we do not grow the number of partitions of live topics.